libbeat/cfgfile/list.go

```diff
@@ -0,0 +1,133 @@
+package cfgfile
+
+import (
+	"sync"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// RunnerList implements a reloadable.List of Runners
+type RunnerList struct {
+	runners  map[uint64]Runner
+	mutex    sync.Mutex
+	factory  RunnerFactory
+	pipeline beat.Pipeline
+}
+
+// NewRunnerList builds and returns a RunnerList
+func NewRunnerList(factory RunnerFactory, pipeline beat.Pipeline) *RunnerList {
+	return &RunnerList{
+		runners:  map[uint64]Runner{},
+		factory:  factory,
+		pipeline: pipeline,
+	}
+}
+
+// Reload the list of runners to match the given state
+// TODO return list of errors (even we don't stop on them)
+func (r *RunnerList) Reload(configs []*common.Config) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	startList := map[uint64]*common.Config{}
+	stopList := r.copyRunnerList()
+
+	debugf("Starting reload procedure, current runners: %d", len(stopList))
+
+	// diff current & desired state, create action lists
+	for _, config := range configs {
+		hash, err := HashConfig(config)
+		if err != nil {
+			logp.Err("Unable to hash given config: %v", err)
+			continue
+		}
+
+		if _, ok := stopList[hash]; ok {
+			delete(stopList, hash)
+		} else {
+			startList[hash] = config
+		}
+	}
+
+	// Stop removed runners
+	for hash, runner := range stopList {
+		debugf("Stopping runner: %s", runner)
+		delete(r.runners, hash)
+		go runner.Stop()
```

Comment:
This is a change in behaviour from previous implementation where it was synchronous. This could become a problem in Filebeat or at least lead to more errors. Before a new prospector can be started on an existing state of a file, the state must be set to Finish. If the shutdown happens in go routine, this overlap can become bigger.  But I think it should still work.